### PR TITLE
Add mapping to ucsc cromosome names for BigWig

### PR DESCRIPTION
**Pull request by JenkoB**
## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [x] All output fields have their ``re-save`` calls.

### CODE REVIEW

**Comment by JenkoB** [comment]
@JureZmrzlikar @mzganec please review

File: resolwe_bio/tests/processes/test_alignment.py [hunk 1]
```diff
@@ -307,3 +309,36 @@ def test_subread(self):
...
+    def test_hisat2_bigwig(self):
+        with self.preparation_stage():
+            paired_reads = self.prepare_paired_reads(mate1=['SRR2124780_1 1k.fastq.gz'],
+                                                     mate2=['SRR2124780_2 1k.fastq.gz'])
+        inputs = {
+            'src': 'hg38_chr21_9M.fa.gz',
+            'species': 'Homo sapiens',
+            'build': 'hg38'
+        }
+        genome_ucsc = self.run_process('upload-genome', inputs)
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_alignment.py]: Since you don't assert fields of `genome_ucsc`, move this into `self.preparation_stage()` context and remove `upload-genome` from the process tag at the top of this test.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 2]
```diff
@@ -307,3 +309,36 @@ def test_subread(self):
...
+
+    @tag_process('alignment-hisat2', 'upload-genome')
+    def test_hisat2_bigwig(self):
+        with self.preparation_stage():
+            paired_reads = self.prepare_paired_reads(mate1=['SRR2124780_1 1k.fastq.gz'],
+                                                     mate2=['SRR2124780_2 1k.fastq.gz'])
+        inputs = {
+            'src': 'hg38_chr21_9M.fa.gz',
+            'species': 'Homo sapiens',
+            'build': 'hg38'
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_alignment.py]: Add a trailing comma.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 3]
```diff
@@ -307,3 +309,36 @@ def test_subread(self):
...
+        }
+        genome_ucsc = self.run_process('upload-genome', inputs)
+        inputs = {
+            'genome': genome_ucsc.id,
+            'reads': paired_reads.id}
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_alignment.py]: Add a trailing comma and Move `}` to the next line.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 4]
```diff
@@ -307,3 +309,36 @@ def test_subread(self):
...
+        aligned_reads = self.run_process('alignment-hisat2', inputs)
+        self.assertFile(aligned_reads, 'bigwig', 'hisat2_paired_ucsc_bigwig.bw')
+
+        with self.preparation_stage():
```
Comment (mzganec) [inline resolwe_bio/tests/processes/test_alignment.py]: It's unusual to have two preparation stages in the same test. The rest of the test starting from here does not have any data objects in common with the first part of the test. Consider adding this preparation stage to the first one or split the test into two test methods.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/python3
+"""Change chromosome names in BigWig header from ensembel or NCBI to UCSC"""
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: ENSEMBL. Make a period at the end to fix linter error.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+import argparse
+import os
+import pyBigWig
+import sys
+
+from os import rename
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Redundant.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/python3
+"""Change chromosome names in BigWig header from ensembel or NCBI to UCSC"""
+
+import argparse
+import os
+import pyBigWig
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Put this import below. System modules first, then everything else.

File: docs/CHANGELOG.rst [hunk 8]
```diff
@@ -27,6 +27,10 @@ Added
   ``alignment-star`` process
 - Add mappings from ENSEMBL or NCBI to UCSC chromosome names to
   ``resolwebio/rnaseq:3.5.0`` docker image
+- Add ``bigwig_chroms_to_ucsc.py`` to ``bamtobigwig.sh`` to change chromosome
```
Comment (JureZmrzlikar) [inline docs/CHANGELOG.rst]: This is just as suggestion, but i would rephrase this to:

Change chromosome names from ENSEMBL / NCBI to UCSC (example: "1" to "chr1") in BigWig files. The purpose of this is to enable viewing BigWig files in UCSC genome browsers for files aligned with ENSEBML or NCBI genome. This change is done by adding script bigwig_chroms_to_ucsc.py to bamtobigwig.sh script.

File: resolwe_bio/tests/processes/test_alignment.py (hunk 2 above)
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_alignment.py]: Add trailing comma.

File: resolwe_bio/tests/processes/test_alignment.py (hunk 3 above)
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_alignment.py]: Add trailing comma and put curly bracket into new line.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 9]
```diff
@@ -307,3 +309,36 @@ def test_subread(self):
...
+            paired_reads = self.prepare_paired_reads(mate1=['GRCh38.p12_NCBI_R1.fq.gz'],
+                                                     mate2=['GRCh38.p12_NCBI_R2.fq.gz'])
+
+        inputs = {
+            'src': 'GRCh38.p12_NCBI_9M.fasta.gz',
+            'species': 'Homo sapiens',
+            'build': 'GRCh38.p12'
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_alignment.py]: Add trailing comma.

File: resolwe_bio/tests/processes/test_alignment.py [hunk 10]
```diff
@@ -307,3 +309,36 @@ def test_subread(self):
...
+        }
+        genome_ncbi = self.run_process('upload-genome', inputs)
+        inputs = {
+            'genome': genome_ncbi.id,
+            'reads': paired_reads.id}
```
Comment (JureZmrzlikar) [inline resolwe_bio/tests/processes/test_alignment.py]: Same as before.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py (hunk 5 above)
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: ENSEMBL. Also, add dot at the end.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py (hunk 7 above)
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Separate standard library imports and third party imports with one blank line.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py (hunk 6 above)
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: This should go under standard library imports.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    ],
+    'Rattus norvegicus': [
+        'rn6_ensembl2UCSC.txt',
+    ],
+}
+
+
+def parse_arguments():
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(description="Convert the chromosome names of a bigWig file.")
```
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Use same description as in the module docstring. You can use ``__doc__``.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def parse_mappings(species, infile, outfile):
+    """Parse file with chromosome mappings."""
+    mappings = dict()
+    # if species doesn't have prepared mapping file the script should exit with status 0 and return BigWig file
+    # with output name and warining
+    if species not in MAPPINGS_FILES:
+        msg = ('Chromosome mappings for Species "{}" are not supported.'.format(species))
+        print(warning(msg))
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Shouldn't we issue an error here? If we exit with `sys.exit(0)`, this won't be caught by `re-checkrc` in `bamtobigwig.sh`.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def parse_mapping_file(filename):
+    """Parse mapping file."""
+    mappings = {}
+    with open(filename) as handle:
+        for line in handle:
+            line = line.strip().split()
+            if len(line) < 2 or line[1] == "":
+                # if line does not conatin names in both sources just skip it
```
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Reword to: ``Skip line, if it does not contain names of both sources``

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    return mappings
+
+
+def parse_mappings(species, infile, outfile):
+    """Parse file with chromosome mappings."""
+    mappings = dict()
+    # if species doesn't have prepared mapping file the script should exit with status 0 and return BigWig file
+    # with output name and warining
+    if species not in MAPPINGS_FILES:
+        msg = ('Chromosome mappings for Species "{}" are not supported.'.format(species))
```
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: You can omit the outer brackets.

File: resolwe_bio/processes/alignment/tophat.yml [hunk 15]
```diff
@@ -252,5 +252,5 @@
       fi
 
       #computation time limit for bamCoverage is 480s per GB
-      bamtobigwig.sh "${NAME}.bam" {{ requirements.resources.cores }} 50 480
+      bamtobigwig.sh "${NAME}.bam" {{ requirements.resources.cores }} 50 480 {{ genome.build }}
```
Comment (mzganec) [inline resolwe_bio/processes/alignment/tophat.yml]: `{{ genome.species }}`

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    for basename in MAPPINGS_FILES[species]:
+        filename = os.path.join(MAPPINGS_DIR, basename)
+        mappings.update(parse_mapping_file(filename))
+    return mappings
+
+
+def create_new_header(infile, mappings, outfile):
+    """Create new header in BigWig, with UCSC chromosome names."""
+    bw = pyBigWig.open(infile)
+    # if chromosome names are UCSC source, then exit with status 0 and return BigWig file with output name
```
Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Reword to: "If chromosome names are already UCSC, just rename input file to output name. Exit with status 0 since this is normal behavior.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    if set(bw.chroms().keys()).issubset(mappings.values()):
+        os.rename(infile, outfile)
+        sys.exit(0)
+    else:
+        hdr = [(mappings[chrom], length) for chrom, length in bw.chroms().items() if chrom in mappings]
+        bwOutput = pyBigWig.open(outfile, "w")
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Is context not supported? `with pyBigWig.open(outfile, 'w') as bwOutput:`

Comment (JureZmrzlikar) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: This variable name does not follow PEP8 recommendation. Rename it.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        bwOutput.addHeader(hdr)
+        for chrom, length in bw.chroms().items():
+            ints = bw.intervals(chrom, 0, length)
+            if len(ints):
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: `if ints:`

File: docs/CHANGELOG.rst (hunk 8 above)
Comment (JenkoB) [reply docs/CHANGELOG.rst]: Nice :D

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py (hunk 12 above)
Comment (JenkoB) [reply resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: That is the plan. The reason is to support the genomes as Dicty, which can not be mapped and showed in UCSC genome browser.

**Comment by JenkoB** [comment]
@mzganec please do another round of review

File: docs/CHANGELOG.rst [hunk 19]
```diff
@@ -13,13 +13,19 @@ Unreleased
...
+  This change is done by adding script bigwig_chroms_to_ucsc.py to
+  bamtobigwig.sh script.
 
 Fixed
 -------
 - fix ``upload-gtf`` to create JBrowse track only if GTF file is ok
 - Pin ``sra-toolkit`` version to 2.9.0 in ``resolwebio/utils`` Docker image.
 
 
+
```
Comment (mzganec) [inline docs/CHANGELOG.rst]: Remove.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+            bw_output.addHeader(hdr)
+            for chrom, length in bw.chroms().items():
+                ints = bw.intervals(chrom, 0, length)
+                if ints:
+                    bw_output.addEntries([mappings[chrom]] * len(ints),
+                                         [x[0] for x in ints],
+                                         ends=[x[1] for x in ints],
+                                         values=[x[2] for x in ints])
+        bw.close()
+        bw_output.close()
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: Redundant due to use of `with` context.

File: resolwe_bio/tools/bigwig_chroms_to_ucsc.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        with pyBigWig.open(outfile, 'w') as bw_output:
+            bw_output.addHeader(hdr)
+            for chrom, length in bw.chroms().items():
+                ints = bw.intervals(chrom, 0, length)
+                if ints:
+                    bw_output.addEntries([mappings[chrom]] * len(ints),
+                                         [x[0] for x in ints],
+                                         ends=[x[1] for x in ints],
+                                         values=[x[2] for x in ints])
+        bw.close()
```
Comment (mzganec) [inline resolwe_bio/tools/bigwig_chroms_to_ucsc.py]: `bw` is not closed if the `else` statement is not executed. Use `with` context as in the case of `bw_output`.
